In [1]:
import pandas as pd
import numpy as np

In [2]:
#download the class movie reviews from here:
#https://docs.google.com/spreadsheets/d/17rCJzmWxqvAu9rkpkgt4ToccIlY4A1Ffuu1W9X3B8Ag/
#then read that in as a pandas dataframe

In [3]:
M = pd.read_csv("Movie Reviews.csv", index_col='Name')

In [4]:
M.head()

American Sniper  The Hunger Games: Mockingjay - Part 1  \
Name                                                                         
Aarti Jaiswal                   4.0                                    NaN   
Aditya Dharmasagar              5.0                                    NaN   
Adrian Cavallaris               NaN                                    4.0   
Ahmed Muheebuddin               5.0                                    NaN   
Aishwarya reddy                 NaN                                    3.0   

                    Guardians of the Galaxy  The Lego Movie  The Hobbit  \
Name                                                                      
Aarti Jaiswal                           NaN             NaN         4.0   
Aditya Dharmasagar                      NaN             4.0         5.0   
Adrian Cavallaris                       3.0             NaN         4.0   
Ahmed Muheebuddin                       4.0             NaN         3.0   
Aishwarya reddy                         4.0             3.0         4.0   

                    Transformers  Malificent  Big Hero 6  Godzilla  \
Name                                                                 
Aarti Jaiswal                3.0         NaN         5.0       NaN   
Aditya Dharmasagar           4.0         4.0         NaN       4.0   
Adrian Cavallaris            2.0         NaN         NaN       NaN   
Ahmed Muheebuddin            NaN         NaN         NaN       NaN   
Aishwarya reddy              4.0         3.0         NaN       4.0   

                    Interstellar  How to Train your Dragon 2  Gone Girl  \
Name                                                                      
Aarti Jaiswal                NaN                         NaN        4.0   
Aditya Dharmasagar           5.0                         5.0        5.0   
Adrian Cavallaris            NaN                         NaN        NaN   
Ahmed Muheebuddin            NaN                         NaN        NaN   
Aishwarya reddy              4.0                         3.0        5.0   

                    Divergent  The Fault in Our Stars  Unbroken  \
Name                                                              
Aarti Jaiswal             NaN                     NaN       NaN   
Aditya Dharmasagar        4.0                     4.0       NaN   
Adrian Cavallaris         4.0                     NaN       NaN   
Ahmed Muheebuddin         NaN                     NaN       NaN   
Aishwarya reddy           NaN                     4.0       NaN   

                    300: Rise of an Empire  
Name                                        
Aarti Jaiswal                          NaN  
Aditya Dharmasagar                     5.0  
Adrian Cavallaris                      NaN  
Ahmed Muheebuddin                      NaN  
Aishwarya reddy                        3.0

In [5]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

In [6]:
def get_recs(movie_name, M, num):

    import numpy as np
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

    

### Question 1:  What movie is most Similar to 'The fault in our stars' (60 pts)

In [7]:
recs = get_recs('The Fault in Our Stars', M, 5)
recs

[('Interstellar', 0.23771516387741332),
 ('Malificent', 0.20294179241370683),
 ('Divergent', 0.16250996411209925),
 ('How to Train your Dragon 2', 0.15744168095198388),
 ('Guardians of the Galaxy', 0.13672667146415826)]


## Q1 Answer: Interstellar with a correlation coefficient of 0.23771516387741332




### Question 2:  Which movie(s) would you most like to see, based on your classmates experience? (40 pts)

In [8]:
#HINTS...

#the movies I've seen
M_copy = M.copy()
my_ratings = M_copy.ix['Daniel Lee'].dropna()
#the movie I've rated the highest
my_ratings.sort(ascending=False)
#the name of the movie I've rated the highest
fav_movie = my_ratings.index[my_ratings == my_ratings[0]]
fav_movie

C:\Users\leeda\Anaconda3\lib\site-packages\ipykernel\__main__.py:7: FutureWarning: sort is deprecated, use sort_values(inplace=True) for INPLACE sorting


Index(['300: Rise of an Empire', 'Gone Girl'], dtype='object')

In [9]:
### Two of my movies are rated 5.0 so run get_recs for both and select highest correlation coefficient

def get_recs3(movie_name, M, num):
    """
    Sames as get_recs but disregard movies I've seen
    """
    import numpy as np
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        if title in my_ratings.index:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

recs2 = get_recs3(fav_movie[0], M, M.shape[1])
recs2 = recs2 + get_recs3(fav_movie[1], M, M.shape[1])
recs2.sort(key=lambda tup: tup[1], reverse=True)
recs2

[('How to Train your Dragon 2', 0.19728895166240468),
 ('American Sniper', 0.17389820131557082),
 ('American Sniper', 0.11862930770262931),
 ('Godzilla', 0.10836186560067539),
 ('The Lego Movie', 0.095720613959421261),
 ('How to Train your Dragon 2', 0.086048673938659781),
 ('Godzilla', 0.083403594188018029),
 ('Unbroken', 0.078167813743160383),
 ('The Fault in Our Stars', 0.070779205719329205),
 ('Unbroken', 0.069253477081581707),
 ('Big Hero 6', -0.00074459180156818754),
 ('The Fault in Our Stars', -0.052487561626630527),
 ('The Lego Movie', -0.10554188888119423),
 ('Big Hero 6', -0.11165814123982593)]

## Q2 Answer: How to Train your Dragon 2 with highest coefficient of correlation of 0.19728895166240468

### Question 3 (Bonus): Ratings for rest of unseen movies

In [10]:
### Create personal ratings vector
my_ratings = M_copy.ix['Daniel Lee']
my_ratings[np.isnan(my_ratings)] = 0
my_ratings

American Sniper                          0.0
The Hunger Games: Mockingjay - Part 1    4.0
Guardians of the Galaxy                  4.0
The Lego Movie                           0.0
The Hobbit                               3.0
Transformers                             3.0
Malificent                               2.0
Big Hero 6                               0.0
Godzilla                                 0.0
Interstellar                             4.0
How to Train your Dragon 2               0.0
Gone Girl                                5.0
Divergent                                3.0
The Fault in Our Stars                   0.0
Unbroken                                 0.0
300: Rise of an Empire                   5.0
Name: Daniel Lee, dtype: float64

In [11]:
### Create similarity matrix  
M_occ = M.copy()
coocc = M_occ.T.dot(M_occ) # matrix shell

# create function to give you correlations
def get_recs2(movie_name, M, num):
    """
    Same as get_recs but 
    1. Sets correlation coefficient to itself as 1 ie perfectly positive correlation
    2. Removes title from results
    3. No post-sorting
    """
    import numpy as np
    
    reviews = []
    for title in M.columns:
        if title == movie_name:
            cor = 1
        else:
            cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append(cor)
    
    #reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

# replace coocc shell with correlations
for x in coocc.columns:
    coocc[x] = get_recs2(x, M, coocc.columns.shape[0])

coocc.head()

American Sniper  \
American Sniper                               1.000000   
The Hunger Games: Mockingjay - Part 1        -0.064664   
Guardians of the Galaxy                      -0.050455   
The Lego Movie                                0.149767   
The Hobbit                                   -0.019540   

                                       The Hunger Games: Mockingjay - Part 1  \
American Sniper                                                    -0.064664   
The Hunger Games: Mockingjay - Part 1                               1.000000   
Guardians of the Galaxy                                             0.231174   
The Lego Movie                                                      0.110598   
The Hobbit                                                          0.182478   

                                       Guardians of the Galaxy  \
American Sniper                                      -0.050455   
The Hunger Games: Mockingjay - Part 1                 0.231174   
Guardians of the Galaxy                               1.000000   
The Lego Movie                                        0.455714   
The Hobbit                                            0.226968   

                                       The Lego Movie  The Hobbit  \
American Sniper                              0.149767   -0.019540   
The Hunger Games: Mockingjay - Part 1        0.110598    0.182478   
Guardians of the Galaxy                      0.455714    0.226968   
The Lego Movie                               1.000000    0.171087   
The Hobbit                                   0.171087    1.000000   

                                       Transformers  Malificent  Big Hero 6  \
American Sniper                            0.041354   -0.208142   -0.102054   
The Hunger Games: Mockingjay - Part 1      0.174304   -0.053693    0.137686   
Guardians of the Galaxy                   -0.057754   -0.059282    0.090330   
The Lego Movie                             0.048239   -0.081631    0.091025   
The Hobbit                                 0.267340    0.147590   -0.013805   

                                       Godzilla  Interstellar  \
American Sniper                       -0.016494      0.141962   
The Hunger Games: Mockingjay - Part 1  0.115324     -0.028089   
Guardians of the Galaxy                0.063978     -0.058907   
The Lego Movie                         0.022157      0.040154   
The Hobbit                             0.092784      0.144440   

                                       How to Train your Dragon 2  Gone Girl  \
American Sniper                                          0.076223   0.173898   
The Hunger Games: Mockingjay - Part 1                   -0.003352  -0.132111   
Guardians of the Galaxy                                  0.137409  -0.042051   
The Lego Movie                                           0.248249   0.095721   
The Hobbit                                               0.066983  -0.145888   

                                       Divergent  The Fault in Our Stars  \
American Sniper                        -0.057356               -0.112809   
The Hunger Games: Mockingjay - Part 1   0.114596                0.040421   
Guardians of the Galaxy                -0.025741                0.136727   
The Lego Movie                         -0.024027                0.045235   
The Hobbit                              0.163974               -0.011598   

                                       Unbroken  300: Rise of an Empire  
American Sniper                       -0.163562                0.118629  
The Hunger Games: Mockingjay - Part 1  0.010990                0.073603  
Guardians of the Galaxy               -0.074491               -0.299337  
The Lego Movie                        -0.143481               -0.105542  
The Hobbit                            -0.178675                0.064161

In [12]:
### Dot product similarity matrix with preference vector
raw_scores = coocc.dot(my_ratings)
raw_scores

American Sniper                          1.047099
The Hunger Games: Mockingjay - Part 1    5.826550
Guardians of the Galaxy                  3.293979
The Lego Movie                           2.799393
The Hobbit                               6.396028
Transformers                             6.557666
Malificent                               3.175381
Big Hero 6                               0.198734
Godzilla                                 4.330372
Interstellar                             5.708707
How to Train your Dragon 2               3.647230
Gone Girl                                3.773004
Divergent                                7.085832
The Fault in Our Stars                   2.537774
Unbroken                                 1.158034
300: Rise of an Empire                   6.196823
dtype: float64

In [13]:
### Normalize with sum of coefficients
predict_ratings = raw_scores.copy()
for x in coocc.index:
    predict_ratings[x] = raw_scores[x]/coocc[x].sum()
predict_ratings.drop(M.ix['Daniel Lee'].dropna().index) # drop movies I've seen before

American Sniper               1.154774
The Lego Movie                1.318438
Big Hero 6                    0.191665
Godzilla                      1.838158
How to Train your Dragon 2    1.347589
The Fault in Our Stars        1.434422
Unbroken                      0.978458
dtype: float64

## Q3 Answer: Ratings listed above.  All are predicted below a rating of 2 which makes sense as I would have seen it already if I wanted to see it or thought it I'd like it.  Best recommendation would be Godzilla, which oddly enough, was a movie I was contemplating seeing.